packages/store/src/external/hydrate.js

```diff
@@ -0,0 +1,113 @@
+// @ts-check
+
+import '../types';
+
+import { makeWeakStore } from '../weak-store';
+import { makeStore } from '../store';
+
+/**
+ * @callback MakeBackingStore
+ * @param {HydrateHook} hydrateHook
+ * @returns {BackingStore}
+ */
+
+/**
+ * This creates an external store maker for a given storage backend, supporting
+ * the Closure interface that the rewriter targets.
+ *
+ * @template {Array<any>} A
+ * @template {ExternalInstance} T
+ * @param {MakeBackingStore} makeBackingStore
+ * @returns {MakeHydrateExternalStore<A, T>}
+ */
+export const makeHydrateExternalStoreMaker = makeBackingStore => {
+  /** @type {WeakStore<T, HydrateKey>} */
+  const instanceToKey = makeWeakStore('instance');
+
+  let lastStoreId = 0;
+
+  // This has to be a strong store, since it is indexed by ID.
+  const storeIdToHydrate = makeStore('storeId');
+
+  /**
+   * Create a data object that queues writes to the store.
+   *
+   * @param {HydrateData} data
+   * @param {() => void} markDirty
+   */
+  const makeActiveData = (data, markDirty) => {
+    const activeData = {};
+    // For every property in data...
+    for (const prop of Object.getOwnPropertyNames(data)) {
+      // Define a getter and setter on activeData.
+      Object.defineProperty(activeData, prop, {
+        get: () => data[prop],
+        set: value => {
+          data[prop] = value;
+          markDirty();
+        },
+      });
+    }
+    return harden(activeData);
+  };
+
+  /** @type {BackingStore} */
+  let backing;
+
+  /** @type {HydrateHook} */
+  const hydrateHook = {
+    getKey(value) {
+      return instanceToKey.get(value);
+    },
+    load([storeId, instanceId]) {
+      const hydrate = storeIdToHydrate.get(storeId);
+      const store = backing.getHydrateStore(storeId);
+
+      const data = store.get(instanceId);
```

Comment:
I'm going to need a walkthrough to be confident that I understand this code, but this fetch has me concerned. From our other discussions about the "representative", we know there will be multiple representatives for the same virtual object, because the code we hand one to might hold on to it for an arbitrary long period of time (and we have no way to tell if/when they've stopped using it). If the data is only fetched from the secondary store once, when we create the representative, it will be stale if/when some other user modifies their own representative.

My hunch is that we must be maximally lazy on read, and have every getter invocation cause a read from secondary storage, to make sure that every `get()` sees the latest value.

Reply:
Yes, you're absolutely right.  I'll make the "active record" something that SwingSet can implement as desired.  That will also solve my problem with the current test (divergence between representatives).